dbms/src/Storages/DeltaMerge/SSTFilesToDTFilesOutputStream.cpp

```diff
@@ -289,6 +307,42 @@ void SSTFilesToDTFilesOutputStream<ChildStream>::cancel()
     }
 }
 
+template <typename ChildStream>
+void SSTFilesToDTFilesOutputStream<ChildStream>::updateRangeFromNonEmptyBlock(Block & block)
+{
+    // TODO: Now we update the rangeEnd for every block. We can update the rangeEnd only once,
+    //  just before a new DTFile is going to be created.
+
+    RUNTIME_CHECK(block.rows() > 0);
+
+    const auto & pk_col = block.getByName(MutableSupport::tidb_pk_column_name);
+    const auto rowkey_column = RowKeyColumnContainer(pk_col.column, schema_snap->is_common_handle);
+    auto & current_file_range = ingest_files_range.back();
+
+    auto const block_start = rowkey_column.getRowKeyValue(0);
+    auto const block_end = rowkey_column.getRowKeyValue(pk_col.column->size() - 1) //
+                               .toRowKeyValue()
+                               .toPrefixNext(); // because range is right-open.
+
+    // TODO: what if there are a lof of versions, exceeding a block size? Will our checks fail?
+    RUNTIME_CHECK(compare(block_start, block_end.toRowKeyValueRef()) < 0);
```

Comment:
For rough set filter working correctly, `BoundedSSTFilesToBlockInputStream` ensures that if there are multiple rows with the same rowkey, they must be in the same block.

But we do need to handle that a block that only contains one rowkey. Calling `toPrefixNext` for getting the block_end may be wrong under this situation.

https://github.com/pingcap/tiflash/blob/b232bcd66ccc2684f9b0688daf941832077affd2/dbms/src/Storages/DeltaMerge/SSTFilesToBlockInputStream.h#L115-L117

Reply:
Interesting, when will we have a block containing only one key? The block size is small, or the versions of key is too much?

Seems that the boundary will be fine, as long as we don't meet one key in multiple blocks.

Consider:

```
[1, 1, 1, 1], [2, ....]
```

The range would be:

```
[1, 2) [2, ...)     (int handle)
[0x1, 0x2) [0x2, ...)  (common handle)
```

Reply:
When the versions of key is too much(8192 rows per block by default), we will get a block containing only one key.

Reply:
I get it, `toPrefixNext` solve the problem